### PR TITLE
K-Core for graphology

### PR DESCRIPTION
**Pull request by 16arpi**
Implementation of the degeneracy calculation already available in the `networkx` python library. It adds to graphology the following methods : 

* coreNumber()
* kCore()
* kShell()
* kCrust()
* kCorona()
* onionLayers()

However, it doesn't include the method `kTruss()`.

The pull request includes tests, a package description and a readme documentation. 

The typescript description file `index.d.ts` needs to be written.

For more details of the original python implementation : [https://networkx.org/documentation/stable/reference/algorithms/core.html](https://networkx.org/documentation/stable/reference/algorithms/core.html)

### CODE REVIEW

File: src/cores/README.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+coreNumber.assign(graph);
+
+// Assign with a custom attribute label
+coreNumber.assign(graph, 'core');
+```
+
+_Arguments_
+
+- **graph** _Graph_: target graph.
+- **coreAttribute** _?string_ : the name of the attribute to use if core numbers are assigned to the nodes.
```
Comment (Yomguithereal) [inline src/cores/README.md:50]: `nodeCoreAttribute` pour être cohérent avec les autres APIs.

File: src/cores/README.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+onionLayers.assign(graph);
+
+// Assign with a custom attribute label
+onionLayers.assign(graph, 'onion');
+```
+
+_Arguments_
+
+- **graph** _Graph_: target graph.
+- **onionLayerAttribute** _?string_ : the name of the attribute to use if onion layers are assigned to the nodes.
```
Comment (Yomguithereal) [inline src/cores/README.md:154]: `nodeOnionLayerAttribute`

File: src/cores/index.js [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+      'graphology-cores: the given graph has self loops which is not permitted.'
+    );
+
+  const degrees = {};
+  graph.forEachNode(node => {
+    degrees[node] = graph.degree(node);
+  });
+
+  // Sort nodes by degree
+  const nodes = Object.entries(degrees)
```
Comment (Yomguithereal) [inline src/cores/index.js:30]: Plutôt trier `graph.nodes()`

File: src/cores/index.js [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    .sort((a, b) => {
+      const d1 = a[1];
+      const d2 = b[1];
+      return d1 - d2;
+    })
+    .map(a => {
+      return a[0];
+    });
+
+  const binBounderie = [0];
```
Comment (Yomguithereal) [inline src/cores/index.js:40]: Peut-être utiliser un array typé de type fixe.

File: src/cores/index.js [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    const nd = nodes[i];
+    nodePos[nd] = i;
+  }
+
+  // Neighbors
+  const nbrs = {};
+  nodes.forEach(node => {
+    nbrs[node] = [];
+    if (graph.type === 'directed') {
+      graph.forEachInboundNeighbor(node, nbr => {
```
Comment (Yomguithereal) [inline src/cores/index.js:67]: `forEachInNeighbor`

File: src/cores/index.js [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+  const removed = new Set();
+  const core = degrees;
+  nodes.forEach(node => {
+    nbrs[node].forEach(nbr => {
+      if (core[nbr] > core[node]) {
+        if (removed.has([node, nbr])) return;
+        removed.add([nbr, node]);
+        const pos = nodePos[nbr];
+        const binStart = binBounderie[core[nbr]];
```
Comment (Yomguithereal) [inline src/cores/index.js:88]: orthographe

File: src/cores/index.js [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        nodes[pos] = stemp;
+
+        binBounderie[core[nbr]] += 1;
+        core[nbr] -= 1;
+      }
+    });
+  });
+
+  if (assign) {
+    graph.forEachNode(node => {
```
Comment (Yomguithereal) [inline src/cores/index.js:103]: Utiliser graph.updateEachNodeAttributes

File: src/cores/index.js [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+  return core;
+}
+
+function coreSubgraph(graph, filter, k, customCore) {
+  if (customCore === undefined) {
+    customCore = coreNumber(false, graph);
+  }
+  if (k === undefined) {
+    k = Math.max.apply(Math, Object.values(customCore));
```
Comment (Yomguithereal) [inline src/cores/index.js:116]: Nope. Ne pas utiliser Object.values et ne pas utiliser apply.

File: src/cores/index.js [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+  }
+  const nodes = [];
+  Object.entries(customCore).forEach(pair => {
```
Comment (Yomguithereal) [inline src/cores/index.js:119]: Ne pas utiliser de callback, faire un for in

File: src/cores/index.js [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+  });
+  return subgraph(graph, nodes);
+}
+
+// ===== Filters =====
+
+function kCore(graph, k, customCore) {
+  return coreSubgraph(
+    graph,
+    (cv, ck, cc) => {
```
Comment (Yomguithereal) [inline src/cores/index.js:133]: Ya des noms plus clairs?

File: src/cores/index.js [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+    customCore
+  );
+}
+
+function kCrust(graph, k, customCore) {
+  if (customCore === undefined) {
+    customCore = coreNumber(false, graph);
+  }
+  if (k === undefined) {
+    k = Math.max.apply(Math, Object.values(customCore)) - 1;
```
Comment (Yomguithereal) [inline src/cores/index.js:157]: Idem

File: src/cores/index.js [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+  }
+  const nodes = [];
+  Object.entries(customCore).forEach(pair => {
```
Comment (Yomguithereal) [inline src/cores/index.js:160]: Idem

File: src/cores/index.js [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+function kCorona(graph, k, customCore) {
+  return coreSubgraph(
+    graph,
+    (cv, ck, cc) => {
+      let nbrsSum = null;
+      if (graph.type === 'directed') {
+        nbrsSum =
+          graph.inboundNeighbors(cv).reduce((acc, nb) => {
```
Comment (Yomguithereal) [inline src/cores/index.js:174]: `inNeighbors`

**Review by Yomguithereal — CHANGES_REQUESTED** [review]

**Comment by AlexisPister** [comment]
Is there something more to do on this PR? I would need to use k-cores on a project using Graphology. Happy to contribute if needed.

**Comment by Yomguithereal** [comment]
@AlexisPister will merge and try to release soon.

@16arpi tu vois j'avais dit que je mergerai un jour :)

**Comment by 16arpi** [comment]
> @16arpi tu vois j'avais dit que je mergerai un jour :)

Je n'ai jamais perdu espoir ahah !

**Comment by Yomguithereal** [comment]
@AlexisPister `graphology-cores` was just published on npm. Docs can be found here: https://graphology.github.io/standard-library/cores

Can you tell me whether this fits your use cases?

**Comment by AlexisPister** [comment]
> @AlexisPister `graphology-cores` was just published on npm. Docs can be found here: https://graphology.github.io/standard-library/cores
> 
> Can you tell me whether this fits your use cases?

Thanks! I'll try this soon and let you know.

**Pull request merged**